File: patterns-use-cases/ticket-reservation/ticket-reservation-python/example/checkout_service.py
import uuid
from typing import TypedDict, List
from restate.context import ObjectContext, Serde
from restate.service import Service

from example.utils.email_client import EmailClient
from example.utils.payment_client import PaymentClient


class Order(TypedDict):
    user_id: str
    tickets: List[str]


payment_client = PaymentClient()
email_client = EmailClient()

checkout = Service("checkout")


@checkout.handler()
async def handle(ctx: ObjectContext, order: Order) -> bool:
    total_price = len(order['tickets']) * 40

    idempotency_key = await ctx.run("idempotency_key", lambda: str(uuid.uuid4()))

    async def pay():
        return await payment_client.call(idempotency_key, total_price)
    success = await ctx.run("payment", pay)

    if success:
        await ctx.run("send_success_email", lambda: email_client.notify_user_of_payment_success(order['user_id']))
    else:
        await ctx.run("send_failure_email", lambda: email_client.notify_user_of_payment_failure(order['user_id']))

    return success

File: tutorials/tour-of-restate-rust/src/part4/checkout_service.rs
use std::collections::HashSet;
use crate::auxiliary::email_client::EmailClient;
use crate::auxiliary::payment_client::PaymentClient;
use restate_sdk::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CheckoutRequest {
    pub(crate) user_id: String,
    pub(crate) tickets: HashSet<String>,
}

#[restate_sdk::service]
pub(crate) trait CheckoutService {
    async fn handle(request: Json<CheckoutRequest>) -> Result<bool, HandlerError>;
}

pub struct CheckoutServiceImpl;

impl CheckoutService for CheckoutServiceImpl {
    // <start_checkout>
    async fn handle(
        &self,
        mut ctx: Context<'_>,
        Json(CheckoutRequest { user_id, tickets }): Json<CheckoutRequest>,
    ) -> Result<bool, HandlerError> {
        let total_price = tickets.len() as f64 * 40.0;

        let idempotency_key = ctx.rand_uuid().to_string();

        let pay_client = PaymentClient::new();
        let success = ctx
            .run(|| PaymentClient.call(&idempotency_key, total_price))
            .await?;

        if success {
            ctx.run(|| EmailClient.notify_user_of_payment_success(&user_id))
                .await?;
        } else {
            ctx.run(|| EmailClient.notify_user_of_payment_failure(&user_id))
                .await?;
        }
        Ok(success)
    }
    // <end_checkout>
}
